hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/GcsEventsHoodieIncrSource.java

```diff
@@ -0,0 +1,228 @@
+  }
+
+  @Override
+  public Pair<Option<Dataset<Row>>, String> fetchNextBatch(Option<String> lastCkptStr, long sourceLimit) {
+    QueryInfo queryInfo = getQueryInfo(lastCkptStr);
+
+    if (queryInfo.areStartAndEndInstantsEqual()) {
+      LOG.info("Already caught up. Begin Checkpoint was: " + queryInfo.getStartInstant());
+      return Pair.of(Option.empty(), queryInfo.getStartInstant());
+    }
+
+    Dataset<Row> sourceForFilenames = queryInfo.initializeSourceForFilenames(srcPath, sparkSession);
+
+    if (sourceForFilenames.isEmpty()) {
+      LOG.info("Source of file names is empty. Returning empty result and endInstant: "
+              + queryInfo.getEndInstant());
+      return Pair.of(Option.empty(), queryInfo.getEndInstant());
+    }
+
+    return extractData(queryInfo, sourceForFilenames);
+  }
+
+  private Pair<Option<Dataset<Row>>, String> extractData(QueryInfo queryInfo, Dataset<Row> sourceForFilenames) {
+    List<String> filepaths = filePathsFetcher.getGcsFilePaths(sparkContext, sourceForFilenames, checkIfFileExists);
+
+    LOG.debug("Extracted " + filepaths.size() + " distinct files."
+            + " Some samples " + filepaths.stream().limit(10).collect(Collectors.toList()));
+
+    Option<Dataset<Row>> fileDataRows = fileDataFetcher.fetchFileData(sparkSession, filepaths, props);
+    return Pair.of(fileDataRows, queryInfo.getEndInstant());
+  }
+
+  private QueryInfo getQueryInfo(Option<String> lastCkptStr) {
+    Option<String> beginInstant = getBeginInstant(lastCkptStr);
+
+    Pair<String, Pair<String, String>> queryInfoPair = calculateBeginAndEndInstants(
+        sparkContext, srcPath, numInstantsPerFetch, beginInstant, missingCheckpointStrategy
+    );
+
+    QueryInfo queryInfo = new QueryInfo(queryInfoPair.getLeft(), queryInfoPair.getRight().getLeft(),
+            queryInfoPair.getRight().getRight());
+
+    if (LOG.isDebugEnabled()) {
+      queryInfo.logDetails();
+    }
+
+    return queryInfo;
+  }
+
+  private Option<String> getBeginInstant(Option<String> lastCheckpoint) {
+    if (lastCheckpoint.isPresent() && !isNullOrEmpty(lastCheckpoint.get())) {
+      return lastCheckpoint;
+    }
+
+    return Option.empty();
+  }
+
+  private static MissingCheckpointStrategy getMissingCheckpointStrategy(TypedProperties props) {
+    boolean readLatestOnMissingCkpt = props.getBoolean(
+            READ_LATEST_INSTANT_ON_MISSING_CKPT, DEFAULT_READ_LATEST_INSTANT_ON_MISSING_CKPT);
+
+    if (readLatestOnMissingCkpt) {
+      return MissingCheckpointStrategy.READ_LATEST;
+    }
+
+    if (props.containsKey(MISSING_CHECKPOINT_STRATEGY)) {
+      return MissingCheckpointStrategy.valueOf(props.getString(MISSING_CHECKPOINT_STRATEGY));
+    }
+
+    return null;
+  }
+
+  /**
+   * Ref: https://hudi.apache.org/docs/gcs_hoodie
+   */
+  private static void addGcsAccessConfs(JavaSparkContext jsc) {
+    jsc.hadoopConfiguration().set("fs.gs.impl", GoogleHadoopFileSystem.class.getName());
```

Comment:
Are these not set already ?

Reply:
No. These are GCS specific configs used only in this module. If they're set, Spark will dynamically instantiate the class given there at a suitable time. 

Having a static reference to the fs.impl is nicer compared to sending the name as a config flag, because this will fail in the constructor of GcsEventsHoodieIncrSource itself if the gcs-connector lib is not present on the classpath! (rather than fail later during the fetch data stage).

Reply:
This will be functional. But, I was just wondering how do other modules that need to access gcs work if it is not already set from outside, eg sink. Do we set this somewhere in S3 incr source too ?

Reply:
Haven't thought about how future GCS modules ought to work.

On the S3 side, the Hudi docs ask you to include aws-java-sdk and hadoop-aws jars - https://hudi.apache.org/docs/s3_hoodie

Within hadoop-aws the S3AFileSystem exists: https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L253-L257

I can't figure out how hadoop-aws lib is brought into the fold when launching the S3 Ingest. The only JAR mentioned on the S3 blog post is aws-java-sdk: https://hudi.apache.org/blog/2021/08/23/s3-events-source/

Reply:
hadoop-aws is provided by the runtime environment, e.g. on EMR.

Reply:
@pramodbiligiri we can actually remove the setting here and document that users need to set all these configs as this doc asks the user anyway.